scipy/special/specfun.c

```diff
@@ -0,0 +1,8418 @@
+        }
+        else {
+            km = (40 + m + n) * (int)(-1.0 - 1.8 * log(x - 1.0));
+        }
+        qf2 = 0.0;
+        qf1 = 1.0;
+        for (k = km; k >= 0; k--) {
+            qf0 = ((2.0 * k + 3.0) * x * qf1 - (k + 2.0 - m) * qf2) / (k + m + 1.0);
+            if (k <= n) {
+                qm[k] = qf0;
+            }
+            qf2 = qf1;
+            qf1 = qf0;
+        }
+        for (k = 0; k <= n; k++) {
+            qm[k] = qm[k] * qm0 / qf0;
+        }
+    }
+
+    if (fabs(x) < 1.0) {
+        for (k = 0; k <= n; k++) {
+            qm[k] = pow(-1, m) * qm[k];
+        }
+    }
+
+    qd[0] = ((1.0 - m) * qm[1] - x * qm[0]) / (x*x - 1.0);
+    for (k = 1; k <= n; k++) {
+        qd[k] = (k * x * qm[k] - (k + m) * qm[k-1]) / (x*x - 1.0);
+    }
+    return;
+}
+
+
+int specfun_msta1(double x, int mp) {
+
+    // ===================================================
+    // Purpose: Determine the starting point for backward
+    //          recurrence such that the magnitude of
+    //          Jn(x) at that point is about 10^(-MP)
+    // Input :  x     --- Argument of Jn(x)
+    //          MP    --- Value of magnitude
+    // Output:  MSTA1 --- Starting point
+    // ===================================================
+
+    int it, nn, n0, n1;
+    double a0, f, f0, f1;
+
+    a0 = fabs(x);
+    n0 = (int)(1.1*a0) + 1;
+    f0 = 0.5*log10(6.28*n0) - n0*log10(1.36*a0/n0)- mp;
+    n1 = n0 + 5;
+    f1 = 0.5*log10(6.28*n1) - n1*log10(1.36*a0/n1) - mp;
+    for (it = 1; it <= 20; it++) {
+        nn = n1 - (n1 - n0) / (1.0 - f0/f1);
+        f = 0.5*log10(6.28*nn) - nn*log10(1.36*a0/nn) - mp;
+        if (abs(nn-n1) < 1) { break; }
+        n0 = n1;
+        f0 = f1;
+        n1 = nn;
+        f1 = f;
+    }
+    return nn;
+}
+
+
+int specfun_msta2(double x, int n, int mp) {
+
+    // ===================================================
+    // Purpose: Determine the starting point for backward
+    //          recurrence such that all Jn(x) has MP
+    //          significant digits
+    // Input :  x  --- Argument of Jn(x)
+    //          n  --- Order of Jn(x)
+    //          MP --- Significant digit
+    // Output:  MSTA2 --- Starting point
+    // ===================================================
+
+    int it, n0, n1, nn;
+    double a0, hmp, ejn, obj, f, f0, f1;
+
+    a0 = fabs(x);
+    hmp = 0.5*mp;
+    ejn = 0.5*log10(6.28*n) - n*log10(1.36*a0/n);
+    if (ejn <= hmp ) {
+        obj = mp;
+        n0 = (int)(1.1*a0) + 1;
+    } else {
+        obj = hmp + ejn;
+        n0 = n;
+    }
+    f0 = 0.5*log10(6.28*n0) - n0*log10(1.36*a0/n0) - obj;
+    n1 = n0 + 5;
+    f1 = 0.5*log10(6.28*n1) - n1*log10(1.36*a0/n1) - obj;
+    for (it = 1; it <= 20; it++) {
+        nn = n1 - (n1 - n0) / (1.0 - f0/f1);
+        f = 0.5*log10(6.28*nn) - nn*log10(1.36*a0/nn) - obj;
+        if (abs(nn-n1) < 1) { break; }
+        n0 = n1;
+        f0 = f1;
+        n1 = nn;
+        f1 = f;
+    }
+    return nn;
```

Comment:
```suggestion
    return nn + 10;
```
https://github.com/scipy/scipy/blob/913c3ffafd1179dccc723a23fca3f0d661f633dc/scipy/special/specfun/specfun.f#L1070-L1071

Reply:
I fixed this one and a few more by carefully making some tests up. However I can't figure out the following problem. In the current scipy release 

```
import scipy.special as ssp
ssp.hyp1f1(0.1, 1, 7.0-24j)
```

gives finite results but if I follow this through which is supposed to hit `cchg` and then `cgama` both fortran and thisC code is giving NaNs. Still can't figure out how is this happening since Fortran is also giving NaNs and wrappers seem to be quite thin.